libs/core/src/main/java/org/opensearch/core/compress/CompressorRegistry.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.core.compress;
+
+import org.opensearch.common.Nullable;
+import org.opensearch.core.common.bytes.BytesReference;
+import org.opensearch.core.compress.spi.CompressorProvider;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+
+import java.io.IOException;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+/**
+ * A registry that wraps a static Map singleton which holds a mapping of unique String names (typically the
+ * compressor header as a string) to registerd {@link Compressor} implementations.
+ *
+ * This enables plugins, modules, extensions to register their own compression implementations through SPI
+ *
+ * @opensearch.internal
+ */
+public final class CompressorRegistry {
+    /** No compression singleton - we still register so users can specify NONE in the API*/
+    public static final Compressor NONE;
+
+    // the backing registry map
+    private static final Map<String, Compressor> registeredCompressors;
+
+    // no instance:
+    private CompressorRegistry() {}
+
+    static {
+        ArrayList<SimpleEntry<String, Compressor>> compressors = new ArrayList<>();
+        for (CompressorProvider provider : ServiceLoader.load(CompressorProvider.class, CompressorProvider.class.getClassLoader())) {
+            compressors.addAll(provider.getCompressors());
+        }
+        registeredCompressors = Map.copyOf(compressors.stream().collect(Collectors.toMap(SimpleEntry::getKey, SimpleEntry::getValue)));
```

Comment:
Nitpick, but I think this can be simplified to:

```suggestion
        registeredCompressors = ServiceLoader.load(CompressorProvider.class, CompressorProvider.class.getClassLoader()).stream()
            .flatMap(p -> p.get().getCompressors().stream())
            .collect(Collectors.toUnmodifiableMap(SimpleEntry::getKey, SimpleEntry::getValue));
```

Can maybe even in-line it in the field definition and skip the static initializer block.